### PR TITLE
ENH Removed history writing in Keras model card

### PR DESCRIPTION
**Pull request by merveenoyan**
This PR solves #869. It's a broken down version of https://github.com/huggingface/huggingface_hub/pull/872.

### CODE REVIEW

**Review by osanseviero — COMMENTED** [review]
Could you please share a repo before and after just to do a diff? Thank you for the PR!

**Comment by merveenoyan** [comment]
Hello, I added a test that checks if history is not written to model card anymore meanwhile saved as a JSON file in the repository. 
Before this PR: https://huggingface.co/merve/model-card-history-removal-before (model pushed with the mixin on main)
After this PR: https://huggingface.co/merve/model-card-history-removal (model pushed with the mixin on this branch)
cc: @adrinjalali @osanseviero

**Review by osanseviero — APPROVED** [review]
Looks nice! I just added a minor suggestion for the test for documenting a bit better.

A bit more generally, this PR removes lots of code that was added relatively recently. This is a sign that maybe we should have spent a bit more time originally on how this would work, since adding the code for the metrics table required quite some time to get right.

**Comment by merveenoyan** [comment]
@osanseviero I agree. I didn't realize a model trained for 100+ epochs would have a super long model card because I'm a bit biased to TF model cards where I observed that the fine-tuning is mostly done in less than 100 epochs 😅

File: tests/test_keras_integration.py [hunk 1]
```diff
@@ -228,6 +228,26 @@ def test_save_pretrained_model_card_fit(self):
         self.assertIn("history.json", files)
         self.assertEqual(len(files), 7)
 
+    def test_save_model_card_history_removal(self):
+        REPO_NAME = repo_name("save")
```
Comment (adrinjalali) [inline tests/test_keras_integration.py:232]: not related to this PR, but `repo_name` seems to be the idea that it's a random repo name, but passing a fixed string makes it less random and creates the randomness from the `time`, which is not really what we want. Not passing any argument here would be a safer option.

File: tests/test_keras_integration.py [hunk 2]
```diff
@@ -228,6 +228,26 @@ def test_save_pretrained_model_card_fit(self):
...
+        model = self.model_init()
+        model = self.model_fit(model)
+
+        save_pretrained_keras(
+            model,
+            f"{WORKING_REPO_DIR}/{REPO_NAME}",
```
Comment (adrinjalali) [inline tests/test_keras_integration.py:238]: this should also be moved to a tempfolder, but not in this PR.

**Review by adrinjalali — COMMENTED** [review]
Thanks for the PR @merveenoyan

File: src/huggingface_hub/keras_mixin.py [hunk 3]
```diff
@@ -223,6 +172,12 @@ def save_pretrained_keras(
         else:
             metadata["tags"] = [task_name]
 
+    if model.history is not None:
```
[the remark this answers is not in the document]
Comment (osanseviero) [reply src/huggingface_hub/keras_mixin.py:175]: What function? `model` is a Keras model as stated in the docstring of this function

**Review by osanseviero — APPROVED** [review]

File: src/huggingface_hub/keras_mixin.py [hunk 4]
```diff
@@ -223,6 +172,12 @@ def save_pretrained_keras(
...
+        if model.history.history != {}:
+            path = os.path.join(save_directory, "history.json")
+            with open(path, "w", encoding="utf-8") as f:
```
[the remark this answers is not in the document]
Comment (merveenoyan) [reply src/huggingface_hub/keras_mixin.py:178]: No we don't 🙂 We take it and put it inside a JSON file.

File: tests/test_keras_integration.py [hunk 5]
```diff
@@ -228,6 +228,26 @@ def test_save_pretrained_model_card_fit(self):
...
+        )
+
+        # Check the history is saved as a json in the repository.
+        files = os.listdir(f"{WORKING_REPO_DIR}/{REPO_NAME}")
```
[the remark this answers is not in the document]
Comment (merveenoyan) [reply tests/test_keras_integration.py:242]: What do you mean?

File: src/huggingface_hub/keras_mixin.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (adrinjalali) [reply src/huggingface_hub/keras_mixin.py:175]: Ah yeah, I looked at the wrong function.

File: tests/test_keras_integration.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (adrinjalali) [reply tests/test_keras_integration.py:242]: As in, having a simple model with a simple history, and check if the history's content is what we expect it to be.

File: src/huggingface_hub/keras_mixin.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (merveenoyan) [reply src/huggingface_hub/keras_mixin.py:178]: People can just see it from the repository without actually loading the model and the `model.history`.

File: tests/test_keras_integration.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (merveenoyan) [reply tests/test_keras_integration.py:242]: sure!

File: src/huggingface_hub/keras_mixin.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (adrinjalali) [reply src/huggingface_hub/keras_mixin.py:178]: This is what I mean:

- the user saves a model, the readme file, and the history file.
- the user at a later time, changes the model (maybe train it a bit more), and saves the history again.

Now the question is, does the second action remove the history saved in the first step? Is that what we want to do? Should we warn the user if we're overwriting an existing history file?

[the remark this answers is not in the document]
Comment (nateraw) [reply src/huggingface_hub/keras_mixin.py:178]: A warning would be nice here, yea. I would say don't break, but warn user. WDYT?

[the remark this answers is not in the document]
Comment (merveenoyan) [reply src/huggingface_hub/keras_mixin.py:178]: SGTM, nice catch @adrinjalali

[the remark this answers is not in the document]
Comment (merveenoyan) [reply src/huggingface_hub/keras_mixin.py:184]: Addressed them here: https://github.com/huggingface/huggingface_hub/pull/876/commits/366460eacf800659264ca9253e5822e0125a7246
It raises a warning and overwrites. Also wrote tests to see if it raises warning and migrated test directories to `tempfile` like you requested.

**Review by nateraw — APPROVED** [review]
LGTM!

I gave it a go here:

- [Colab](https://gist.github.com/nateraw/1b516a3eeb89fac6eca7e058b1b22a23)
- [Model Repo](https://huggingface.co/nateraw/autoencoder-keras-rm-history-pr-review)

File: tests/test_keras_integration.py [hunk 6]
```diff
@@ -226,8 +227,49 @@ def test_save_pretrained_model_card_fit(self):
...
+            files = os.listdir(f"{tmpdirname}/{WORKING_REPO_DIR}/{REPO_NAME}")
+            self.assertIn("history.json", files)
+
+            # Check that there is no "Training Metrics" section in the model card.
+            # This was done in an older version.
+            with open(
+                f"{tmpdirname}/{WORKING_REPO_DIR}/{REPO_NAME}/README.md", "r"
+            ) as file:
+                data = file.read()
+            self.assertNotIn(data, "Training Metrics")
```
[the remark this answers is not in the document]
Comment (merveenoyan) [reply tests/test_keras_integration.py:271]: it's the section we got rid of in the model card.

File: src/huggingface_hub/keras_mixin.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (adrinjalali) [reply src/huggingface_hub/keras_mixin.py:184]: If the user hasn't tracked the history in git already, this means this method erases the existing history and the user doesn't really have a way to recover that. I think as a user that can come as a surprise. Also, warnings usually should have a way to be silenced in a reasonable way, but here the only way to silence the warning is to delete that file before calling this method and that doesn't sound like good practice to me.

If we really don't want to append to the history, we can take a backup of the existing file (making sure we don't overwrite another backup it that one exists), warn, and then overwrite the existing one.

If y'all think these concerns are not valid and users don't care about it, then happy to have this merged. But I do think it's a surprising behavior for users.

**Pull request merged**